docs/contributing/api for key modules using typedoc

bra1nDump opened the pull request:
**Draft** for issue #415.
> Issue suggests to add the following to contributing docs:
> 1. Architecture overview
> 2. Links to key functions
> 3. Static API documentation from inline docs

Originally I thought generating documentation for all files might be overwhelming and unnecessary so I did a partial generation for key files. This didn't work out great - many types became un clickable. 

Current approach
* Serve generated docs for full API from gh-pages. To avoid commit noise.
* Create architecture overview in markdown as part of the repo with links to static docs. My initial stab at it is in `CommandRunner.runCommand`. Combined with the video code walk through from Pokey this should be a good starting point.
* Serve the overview as a home page for the docs. By itself a reference would probably not be useful, but combined with an overview ideally one could stay on the documentation website and get a good sense of the project internals.

bra1nDump commented:
A bunch of pushes while debugging github pages action. Turns out the action simply pushed to gh-pages branch. To host the page additional setup was needdd in settings for the repo :D

pokey commented on .github/workflows/publish-docs.yml:
```diff
@@ -0,0 +1,31 @@
+      - name: Check out
+        uses: actions/checkout@v2
+
+      - uses: borales/actions-yarn@v2.3.0
+        with:
+          cmd: install
+
+      - uses: borales/actions-yarn@v2.3.0
+        with:
+          cmd: docs-build
```
`borales/actions-yarn` seems to be somewhat deprecated.  why not just use `yarn` on cli like we do for unit tests? https://github.com/pokey/cursorless-vscode/blob/715a8e77fb41ffd413f5349a8634061f117ec2db/.github/workflows/test.yml#L21-L25

pokey commented on package.json:
```diff
@@ -426,10 +426,11 @@
 	},
 	"scripts": {
-		"vscode:prepublish": "npm run -S esbuild-base -- --minify",
+		"vscode:prepublish": "npm run -S esbuild-base -- --minify && npm run docs-compile",
```
how come we run this one in `prepublish`?

pokey commented on src/core/commandRunner/CommandRunner.ts:
```diff
@@ -17,6 +17,9 @@ import { isString } from "../../util/type";
   private disposables: vscode.Disposable[] = [];
 
+  /**
+   * Subscribe to vscode `cursorless.command`.
```
Not sure I follow this comment

pokey commented on typedoc.json:
```diff
@@ -0,0 +1,20 @@
+{
+    "entryPoints": [
+        "src/extension.ts",
+        "src/core/commandRunner/CommandRunner.ts",
+        "src/core/inferFullTargets.ts",
+        "src/core/updateSelections/RangeUpdater.ts",
+        "src/core/updateSelections/updateRangeInfos.ts",
+        "src/core/updateSelections/updateSelections.ts",
+        "src/processTargets/index.ts",
+        "src/processTargets/modifiers/processModifier.ts",
+        "src/processTargets/processMark.ts",
+        "src/typings/Types.ts",
+        "src/typings/treeSitter.ts",
+        "src/util/canonicalizeAndValidateCommand.ts",
+        "src/util/getPrimitiveTargets.ts"
```
how did you choose these entrypoints?  Why not just everything?

bra1nDump replied:
Thats is what I will do now. For some reason I thought to only create docs for 'important modules'. And yes they were hand picked by me

pokey commented on typedoc.json:
```diff
@@ -0,0 +1,20 @@
+        "src/util/canonicalizeAndValidateCommand.ts",
+        "src/util/getPrimitiveTargets.ts"
+    ],
+    "readme": "none", 
```
so we can set this so that we get a better landing page?

bra1nDump replied:
Yes! That is the plan. Especially if we generate all apis it will be overwhelming without pointing to the key classes and functions.

pokey commented on .github/workflows/publish-docs.yml:
```diff
@@ -0,0 +1,31 @@
+  push:
+    branches:
+      # NOTE: You may want to limit the trigger branch to be "main" or "master" etc.
+      - '*'
```
yeah we prob just want this one on `main` right?  otherwise every PR will clobber our docs

bra1nDump replied:
Also the plan. I did this for testing github action on my fork.

pokey commented on .github/workflows/publish-docs.yml:
```diff
@@ -0,0 +1,31 @@
+
+      - name: Deploy to GitHub Pages
+        if: success()
+        uses: crazy-max/ghaction-github-pages@v2
```
how'd you land on this one?  [This one](https://github.com/peaceiris/actions-gh-pages) for instance has 10x as many stars, so maybe let's use that one?

pokey commented on typedoc.json:
```diff
@@ -0,0 +1,20 @@
+    ],
+    "readme": "none", 
+    "excludePrivate": true,
+    "out": "docs/contributing/api"
```
Given we are just doing this transiently, maybe let's put it in a dir that makes it look more auto-generated, and then add it to `.gitignore`?  Maybe `build/docs/api`?  I was thinking `docs/contributing/api` only if we were actually planning to commit the docs to source

pokey replied:
Although on second thought, maybe we want to just publish the whole `docs` dir?  In that case I think let's stick with this directory, but add it to `.gitignore`

pokey reviewed (commented):
Great start thanks for putting up this draft!  Typedoc seems pretty cool; I could def see these docs being really helpful for contributors, and it's great to have a place to link to from architecture notes, etc.

I left a few comments.  Also, when I ran this one locally, I got a bunch of error output, which seemed to result in a bunch of broken links.  Any idea what might have caused this?

```
➜ yarn docs-build
yarn run v1.22.17
$ npx typedoc
Warning: FullSelectionInfo, defined at src/typings/updateSelections.ts:72, is referenced by core/updateSelections/updateSelections.getSelectionInfo.getSelectionInfo but not included in the documentation.
Warning: HatStyleName, defined at src/core/constants.ts:29, is referenced by typings/Types.DecoratedSymbol.symbolColor but not included in the documentation.
Warning: CommandServerApi, defined at src/util/getExtensionApi.ts:24, is referenced by typings/Types.Graph.commandServerApi but not included in the documentation.
Warning: default, defined at src/core/Debug.ts:11, is referenced by typings/Types.Graph.debug but not included in the documentation.
Warning: default, defined at src/core/Decorations.ts:35, is referenced by typings/Types.Graph.decorations but not included in the documentation.
Warning: EditStyles, defined at src/core/editStyles.ts:41, is referenced by typings/Types.Graph.editStyles but not included in the documentation.
Warning: default, defined at src/core/FontMeasurements.ts:6, is referenced by typings/Types.Graph.fontMeasurements but not included in the documentation.
Warning: default, defined at src/core/HatTokenMap.ts:15, is referenced by typings/Types.Graph.hatTokenMap but not included in the documentation.
Warning: Snippets, defined at src/core/Snippets.ts:16, is referenced by typings/Types.Graph.snippets but not included in the documentation.
Warning: ReadOnlyHatMap, defined at src/core/IndividualHatMap.ts:5, is referenced by typings/Types.ProcessedTargetsContext.hatTokenMap but not included in the documentation.
Warning: ExpansionBehavior, defined at src/typings/updateSelections.ts:24, is referenced by typings/Types.Token.expansionBehavior but not included in the documentation.
Warning: RangeOffsets, defined at src/typings/updateSelections.ts:6, is referenced by typings/Types.Token.offsets but not included in the documentation.
Warning: CommandArgumentComplete, defined at src/core/commandRunner/types.ts:2, is referenced by util/canonicalizeAndValidateCommand.canonicalizeAndValidateCommand.canonicalizeAndValidateCommand but not included in the documentation.
Warning: ExtendedTextDocumentChangeEvent, defined at src/typings/updateSelections.ts:47, is referenced by core/updateSelections/updateRangeInfos.updateRangeInfos.updateRangeInfos.changeEvent but not included in the documentation.
Warning: FullRangeInfo, defined at src/typings/updateSelections.ts:34, is referenced by core/updateSelections/updateRangeInfos.updateRangeInfos.updateRangeInfos.rangeInfoGenerator but not included in the documentation.
Warning: SelectionInfo, defined at src/typings/updateSelections.ts:68, is referenced by core/updateSelections/updateSelections.performEditsAndUpdateSelectionInfos.performEditsAndUpdateSelectionInfos.originalSelectionInfos but not included in the documentation.
Warning: CommandArgument, defined at src/core/commandRunner/types.ts:8, is referenced by util/canonicalizeAndValidateCommand.canonicalizeAndValidateCommand.canonicalizeAndValidateCommand.commandArgument but not included in the documentation.
Warning: ThatMark, defined at src/core/ThatMark.ts:2, is referenced by core/commandRunner/CommandRunner.default.constructor.new default.thatMark but not included in the documentation.
Warning: TestCaseRecorder, defined at src/testUtil/TestCaseRecorder.ts:7, is referenced by core/commandRunner/CommandRunner.default.constructor.new default.testCaseRecorder but not included in the documentation.
Warning: SnippetMap, defined at src/typings/snippet.ts:48, is referenced by extension.activate.activate.__type.__type.__type.__type.snippets but not included in the documentation.
Info: Documentation generated at ./docs/contributing/api
```

bra1nDump commented:
Oh that is caused by generating partial APIs. Types are referenced in type signatures that are not included in the API. This results in non clickable types :( I will fix that.

bra1nDump commented:
* Addressed most comments. Readme in the next commit.
* Default exports switched to named for typedoc to be able to reference using class names. Also final documentation looks better when a class isn't simply called `default`.

I have looked for a way to migrate default exports to named ones, but the best that I found is migrating one by one using inline hints in visual studio code. Unfortunately seems to only be doable one by one. I suspect some linter can do this.

pokey reviewed (commented):
Def looking better

bra1nDump commented on .github/workflows/publish-docs.yml:
```diff
@@ -0,0 +1,24 @@
+        uses: peaceiris/actions-gh-pages@v3
+        with:
+          github_token: ${{ secrets.GITHUB_TOKEN }}
+          publish_dir: ./docs/contributing/api
```
I am thinking how publishing the whole directory would work. Would like any suggestions on this.

My current idea is
* Typedoc generates markdown API (like originally planned).
* Copy readme and contributing to docs/.
* Use https://docsify.js.org/#/ to serve project readme as root and the rest of docs/. 

This way we will have a single page to access all info needed to use and/or get started with contributing.

bra1nDump commented:
Notes from the meetup:
* All contributers are on VSCode, maybe there is a way to use some VSCode extensions?
* How are common npm packages hosting their docs?

Resources
* Typedoc -> sphinx https://github.com/mozilla/sphinx-js
* https://blog.cloudflare.com/generating-documentation-for-typescript-projects/
* https://dotnet.github.io/docfx/
* https://tsdoc.org/

pokey commented:
> What we want:
> 
> * Permanent links. Once we publish it, it doesn't break.
> 
> * Architecture overview.
> 
> * Linking from one function to another.
> 
> * Search.
> 
> 
> 
> Notes from the meetup:
> 
> * All contributers are on VSCode, maybe there is a way to use some VSCode extensions?
> 
> * How are common npm packages hosting their docs?
> 
> * Look into how rust docs
> 
> 
> 
> Resources
> 
> * https://tsdoc.org/
> * [DocFX + TypeDoc](https://dotnet.Github.io/docfx/tutorial/universalreference/gen_doc_for_ts.html)
> 
> * Typedoc -> sphinx https://github.com/mozilla/sphinx-js

Thanks for capturing this! Looks good to me

pokey commented:
How come there are still two open PRs here? Shall we close one?

bra1nDump commented:
Oh no I thought it was just a test branch. I was using it to test deployment to gh pages, forgot that its linked to this PR..
